lib/http2/debug_state.c

```diff
@@ -0,0 +1,217 @@
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include <inttypes.h>
+#include "h2o.h"
+#include "h2o/http2.h"
+#include "h2o/http2_internal.h"
+
+static const char *debug_state_string_open = "OPEN";
+static const char *debug_state_string_half_closed_remote = "HALF_CLOSED_REMOTE";
+static const char *debug_state_string_reserved_local = "RESERVED_LOCAL";
+
+const char *get_debug_state_string(h2o_http2_stream_t *stream)
+{
+    if (h2o_http2_stream_is_push(stream->stream_id)) {
+        switch (stream->state) {
+        case H2O_HTTP2_STREAM_STATE_RECV_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+            return debug_state_string_reserved_local;
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    } else {
+        switch (stream->state) {
+        case H2O_HTTP2_STREAM_STATE_RECV_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+            return debug_state_string_open;
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+static void append_line(h2o_mem_pool_t *pool, h2o_iovec_vector_t *lines, const char *fmt, ...)
+{
+    char check[1]; /* for only checking the size of output */
+    va_list args;
+
+    va_start(args, fmt);
+    int size = vsnprintf(check, 1, fmt, args);
+    va_end(args);
+
+    h2o_iovec_t v;
+    v.base = h2o_mem_alloc_pool(pool, size + 1);
+
+    va_start(args, fmt);
```

Comment:
Please use `va_copy`. Applying `va_start` more than once against one `va_list` is not permitted (IIRC the behavior is undefined).